build/helper/metadata_add_all.py

```diff
@@ -439,39 +440,56 @@ def add_all_attribute_metadata(attributes, config):
 
 
 def _add_enum_codegen_method(enums, config):
-    '''Adds 'codegen_method' that will determine whether and how the enum is code genned. Default is public
+    '''Adds 'codegen_method' that will determine whether and how the enum is code-generated.
```

Comment:
I think the rule is that enums default to not getting generated unless they are used by a function or attribute, in which case 'codegen_method' is set to match to that of whatever uses it.

Reply:
If what I said above is correct, then you can hopefully this unreadable function...

you can do it in term of functions like:

`def get_functions_that_use_enum(enum_name)`
`def get_attributes_that_use_enum(attribute_name)`
`def get_least_restrictive_codegen_method(codegen_method`

You can add a unit test

Reply:
> It adds codegen method only if not explictly specified in metadata.

Not really, it might change an explicitly specified `codegen_method` of  'no' to 'private' if there is a private method / attribute using it.

> I think the rule is that enums default to not getting generated unless they are used by a function or attribute, in which case 'codegen_method' is set to match to that of whatever uses it.

Yes, this is right. I believe this is what my docstring is trying to explain: 
```
    Set the 'codegen_method' of all enums that do not explicitly specify it in metadata to 'no',
    then go through all functions and attributes and set to the least restrictive use.
```

> you can do it in term of functions like:
>
> def get_functions_that_use_enum(enum_name)
> def get_attributes_that_use_enum(attribute_name)
> def get_least_restrictive_codegen_method(codegen_method

I could not really find a way to break this function further down to your suggested functions as they are really intertwined, e.g. `get_least_restrictive_codegen_method()` would have to iterate through all the functions and attributes before arriving at the final results, which is essentially what this `_add_enum_codegen_method()` function is doing.

Reply:
> Not really, it might change an explicitly specified codegen_method of 'no' to 'private' if there is a private method / attribute using it.

So the codegenerator will override the explicit instructions from metadata? That's not a good idea. It should instead error out with a good message, something like:

> Enum `Colors` specifies `'codegen_method': 'no'`, but attribute `color` uses it and has `'codegen_method': 'private'` which is incompatible. Please fix your metadata.

Reply:
> Yes, this is right. I believe this is what my docstring is trying to explain:

Well, the comment describes the algorithm, not what it does at a high level which is more helpful.

Reply:
> I could not really find a way to break this function further down to your suggested functions as they are really intertwined,

Maybe once we stop overriding explicit metadata this is easier to do?

Reply:
This is being resolved by #1767.